GitHub actions CI improvements

TimothyWillard opened the pull request:
### Describe your changes.

This pull request:

- Switches the GitHub actions CI to use a unified conda install done via the `./bin/flepimop-install-dev` script which closely replicates user installation and enables cache sharing across GitHub action workflows, #475.
- Enables doctest for `gempyor`, #563.
- Also upped the min pylint score from 4 to 5, the latest score from CI was a 7.08 (https://github.com/HopkinsIDD/flepiMoP/actions/runs/16377836820/job/46282026928?pr=581).
- Updated the formatting of the produced `environment.yml` file to be more friendly for IDEs, #485.
- Removed the push trigger on the `dev` branch, in theory sounded like a good idea but in practice results in actions running twice when we have a release PR open.
- Made the slow unit tests for `gempyor` run as a dependent job of the fast tests and only run with one python version.
- Spun out R package installer script into a standalone script for modularization and to assist with R package development.

### Does this pull request make any user interface changes? If so please describe.

Those are reflected in updates to the documentation in `python-guidelines-for-developers.md`.

### What does your pull request address? Tag relevant issues.

This pull request addresses #475, #563.

TimothyWillard committed:
Use conda and install scripts in CI workflows

Created a custom `flepimop-install` action and incorporated it into the
gempyor, flepicommon, and inference CI workflows. This custom action
takes care of:
* Installing apt packages required by some of the R dependencies,
* Installing miniconda,
* Either pulling a conda venv from cache or installing flepiMoP with the
  standard install scripts with a user specified python/R version, and
* Displaying install information for debugging purposes.

TimothyWillard committed:
Pull flepiMoP install info out of install

Pulled the script that generates info on the flepiMoP install out of
`bin/flepimop-install` into its own `bin/flepimop-install-info` for
standalone use.

Also updated installation script to also install the tests for the R
packages.

TimothyWillard committed:
Update `environment.yml` via GitHub action

TimothyWillard committed:
Add doctests to gempyor CI

Added a step to the gempyor CI workflow to run doctests as well as fixed
the current state of doctests throughout the gempyor package. Mostly
quick fixes, more detailed and higher quality examples will be required
in follow ups.

TimothyWillard committed:
Improve `environment.yml` file formatting

Add spacing to list item entries so IDEs are a bit less confused when
viewing the file. Addresses #485.

TimothyWillard committed:
Update `environment.yml` via GitHub action

emprzy commented on flepimop/gempyor_pkg/src/gempyor/distributions.py:
```diff
@@ -443,13 +444,13 @@ class BetaDistribution(DistributionABC):
-        array([0.31153835])
+        array([0.24395464])
         >>> dist.sample(size=(3, 5), rng=rng)
-        array([[0.18343469, 0.43545995, 0.48512838, 0.08625927, 0.22223849],
```
How come you changed the numbers in the docstring examples? Just curious

TimothyWillard replied:
Doctests were failing without the change, I suspect the original example was generated with an RNG of a different seed or something along those lines.

emprzy replied:
Hm, interesting.

TimothyWillard committed:
Remove push to `dev` trigger, split out slow tests

* Removed the trigger whenever there is a push to the `dev` branch for
	the GitHub actions, in practice results in actions being ran twice on
  PR from `dev` into `main`.
* Split the slow gempyor tests out into their own job that depends on
	the fast tests job and only runs with python 3.11. Running the slow
  tests with all python versions is not needed.

TimothyWillard committed:
Extract out R pkg installer

Extracted out the custom R pkg installer into its own
`bin/flepimop-install-r-pkgs` script. This is helpful, because:
1. It helps to modularize `bin/flepimop-install` a little bit more for
	 maintainability, and
2. Assists with development of R packages by making them easier to
	 reinstall for testing. R does not have an `--editable` install like
   python does so edits to `flepimop-inference-*` require reinstall.

jc-macdonald reviewed (approved):
Looks good to me, nice work @TimothyWillard

pearsonca reviewed (approved):
lgtm
